Plugins/NativeXr/Source/NativeXr.cpp

```diff
@@ -2003,19 +2040,34 @@ namespace Babylon
 
                     current.insert(inputSource.ID);
 
-                    auto found = m_idToInputSource.find(inputSource.ID);
-                    if (found == m_idToInputSource.end())
+                    auto inputSourceFound = m_idToInputSource.find(inputSource.ID);
+                    if (inputSourceFound == m_idToInputSource.end())
                     {
                         // Create the new input source, which will have the correct spaces associated with it.
                         m_idToInputSource.insert({inputSource.ID, CreateXRInputSource(inputSource, env)});
+                        
+                        //Now that input Source is created, create a gamepad object if enabled for the input source
+                        inputSourceFound = m_idToInputSource.find(inputSource.ID);
+                        if (inputSource.GamepadTrackedThisFrame)
+                        {
+                            auto inputSourceVal = inputSourceFound->second.Value();
+                            CreateXRGamepadObject(inputSourceVal, inputSource);
+                        }
 
                         added.insert(inputSource.ID);
                     }
                     else
                     {
                         // Ensure the correct spaces are associated with the existing input source.
-                        auto val = found->second.Value();
-                        SetXRInputSourceData(val, inputSource);
+                        auto inputSourceVal = inputSourceFound->second.Value();
+                        SetXRInputSourceData(inputSourceVal, inputSource);
+
+                        //inputSource already exists, find the corresponding gamepad object if enabled and set to correct values
+                        if (inputSourceVal.Has("gamepad"))
+                        {
+                            auto gamepadObject = inputSourceVal.Get("gamepad").As<Napi::Object>();
```

Comment:
If you wanted to, you could now move this line inside of SetXRGamepadObject, or heck even put this whole if block inside of SetXRInputSourceData

Reply:
So it seems like I can do that from the surface but its not possible. SetXRGamepadObject is also called from CreateXRGamepadObject before the gamepad is even set. So, here is where the gamepad property is set to the inputsource.  SetXRGamepadObject has to be called every other time with already created gamepad except for the first time  . If I move the gamepad extraction into the SetXRGamepadObject, it won't work the first time.

Reply:
Oh, I missed that part.